File: RustTracer/Raytracer/vector.rs
pub struct Vector {
    pub x: f64,
    pub y: f64,
    pub z: f64
}

impl Vector {
	pub fn dot(&self, v:Vector) -> f64 {
		(self.x * v.x) + (self.y * v.y) + (self.z * v.z)
	} 

	pub fn length(&self) -> f64 {
		self.dot(*self).sqrt()
	}

	pub fn cross(&self, v:Vector) -> Vector {
		Vector {x: self.y * v.z - self.z * v.y,
				y: self.z * v.x - self.x * v.z,
				z: self.x * v.y - self.y * v.x}
	}

	pub fn add(&self, v:Vector) -> Vector {
		Vector {x: self.x + v.x, y: self.y + v.y, z: self.z + v.z }
	}

	pub fn subtract(&self, v:Vector) -> Vector {
		Vector {x: self.x - v.x, y: self.y - v.y, z: self.z - v.z }
	}

	pub fn scale(&self, s:f64) -> Vector {
		Vector {x: self.x * s, y: self.y * s, z: self.z * s }
	}

	pub fn normalize(&self) -> Vector {
		let mag = self.length();
		self.scale(if mag == 0.0 { Float::infinity() } else { 1.0 / mag })
	}
}

File: CPlusPlusTracer/CPlusPlusTracer/Vector.cpp
#include "Vector.h"

CVector::CVector(double x, double y, double z) : X(x), Y(y), Z(z) {
}

double CVector::Dot(CVector v) const {
   return (this->X * v.X) + (this->Y * v.Y) + (this->Z * v.Z);
}

double CVector::Length() const {
   return sqrt(this->Dot(*this));
}

CVector CVector::Cross(CVector v) const {
   return
      CVector((this->Y*v.Z) - (this->Z * v.Y),
	          (this->Z*v.X) - (this->X * v.Z),
              (this->X*v.Y) - (this->Y * v.X));
}

CVector CVector::Add(CVector v) const {
   return CVector(this->X + v.X, this->Y + v.Y, this->Z + v.Z);
}

CVector CVector::Subtract(CVector v) const {
   return CVector(this->X - v.X, this->Y - v.Y, this->Z - v.Z);
}

CVector CVector::Scale(double s) const {
   return CVector(this->X * s, this->Y * s, this->Z * s);
}

CVector CVector::Normalize() const {
   double mag = this->Length();
   return this->Scale(mag == 0.0 ? INFINITY : (1.0 / mag));
}

double CVector::GetX() const {
   return this->X;
}

double  CVector::GetY() const {
   return this->Y;
}

double  CVector::GetZ() const {
   return this->Z;
}